daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/UDFunctions.scala

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dpath
+
+case class UDFunctionCall(recipes: List[CompiledDPath], functionClassObject: Object)
+  extends FNArgsList(recipes) {
+
+  lazy val maybeEvaluate = functionClassObject.getClass.getMethods.collect { case p if p.getName == "evaluate" => p }
```

Comment:
So this getting of the evaluate method won't throw because we checked for it elsewhere already and so we know it is here. That's worth an Assert.invariant(maybeEvaluate.isDefined), and a comment that we already know it is present. 

That said, since we had to check for it earlier, why don't we save it in the functionClassObject earlier, and just access it here? i.e., functionClassObject.evaluateMethod ?

Reply:
Unfortunately Reflection Methods are not serializable, so we got an error (from writeObject) at the final stage of the Compile when we tried to return the evaluate method in the expression, hence the duplication.

I'll put in an assert invariant though

Reply:
...to check that maybeEvaluate is not empty and is not more that 1. I'll also change the name to better reflect that it's a list of possible evaluate methods.

Reply:
Also please add your comment above about the non-serializability to the code as a comment, because otherwise I'd refactor this, and waste my time running into this thing you already ran into and figured out.

Reply:
We might want to tweak how this all works in that case. We probably want to make this non-lazy or explicitly call it so that it's calculated at schema compile time so we don't take that reflection hit on lookup. But then make it transient and add some deserialization logic so that we relookup the eval method on deserializtion. This way, we only take the hit on schema compliation time and on reload, but never at runtime.

Reply:
I retested this and turns out I was mistaken. We are able to serialize both UserDefinedFunction and Methods with no issues, so we only do the reflection lookup for the evaluate method once